qiskit_experiments/framework/json.py

```diff
@@ -477,6 +477,14 @@ def default(self, obj: Any) -> Any:  # pylint: disable=arguments-differ
                     "version": get_object_version(cls),
                 },
             }
+        if isinstance(obj, Gate):
+            # Serialize  gate by storing it in a circuit.
+            circuit = QuantumCircuit(obj.num_qubits)
+            circuit.append(obj, range(obj.num_qubits))
+            value = _serialize_and_encode(
+                data=circuit, serializer=lambda buff, data: qpy_serialization.dump(data, buff)
+            )
+            return {"__type__": "Gate", "__value__": value}
```

Comment:
What happens if you just set circuit here? It is not recursively serialized?

Reply:
It should be able to do it recursively if you set the circuit here, so either way would be fine.

We should add direct serialization unit tests for custom user subclasses of Instructions and Gates, along with tests for opaque pulse gates.

Reply:
I added a few more tests here [ff1ab0f](https://github.com/Qiskit/qiskit-experiments/pull/684/commits/ff1ab0f83638656c63898073093d2eeb28e19174). Note that this does not seem to serialize properly:
```python
class CustomInstruction(Instruction):
    """A custom instruction for testing."""

    def __init__(self):
        """Initialize the instruction."""
        super().__init__("test_inst", 123, 5, [0.45, 0.6])
```
Should we support something like that? @chriseclectic

Reply:
The above works after [1bc6005](https://github.com/Qiskit/qiskit-experiments/pull/684/commits/1bc6005360ec608d244d28ba7b29d9d8860a24c7) it was missing the classical bits.